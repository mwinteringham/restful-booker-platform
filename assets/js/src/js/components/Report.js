import React, { useEffect, useState } from 'react';
import { Calendar, momentLocalizer} from 'react-big-calendar';
import moment from 'moment/moment';
import AdminBooking from './AdminBooking.js';

import "react-big-calendar/lib/css/react-big-calendar.css";
import { API } from '../libs/Api';

const Report = ({defaultDate}) => {

  const [report, setReport] = useState([]);
  const [showBookingForm, toggleBookingForm] = useState(false);
  const [dates, setDates] = useState(null);

  useEffect(() => {
    API.getReport(setReport);
  }, [])

  const addBooking = (result) => {
    if(result.slots.length > 1){
      setDates(result);
      toggleBookingForm(true);
    }
  }

  const closeBooking = () => {
    toggleBookingForm(false);

    API.getReport(setReport);
  }

  const localizer = momentLocalizer(moment);

  if(showBookingForm){
    return <AdminBooking closeBooking={closeBooking} dates={dates} />
  } else {
    return <div>
      <Calendar
        defaultDate={defaultDate}
        onSelectSlot={addBooking}
        selectable
        localizer={localizer}
        defaultView="month"
        popup={true}
        events={report}
        style={{ height: "75vh" }}
        views={['month']}
      />
    </div>
  }
}

export default Report;
