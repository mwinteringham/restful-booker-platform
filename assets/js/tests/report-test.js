import React from 'react';
import Report from '../src/js/components/Report.js';
import nock from 'nock';
import '@testing-library/jest-dom'
import { screen } from '@testing-library/react'


import {
  render, waitFor
} from '@testing-library/react'

test('Multiple reports can be created in the Report component', async () => {
    const mockReport = nock('http://localhost')
                        .get('/report/')
                        .reply(200,  {
                          report : [
                            {
                              start: "2019-04-01",
                              end: "2019-04-03",
                              title: "101"
                            },{
                              start: "2019-04-02",
                              end: "2019-04-04",
                              title: "102"
                            },{
                              start: "2019-04-02",
                              end: "2019-04-04",
                              title: "103"
                            }
                          ]
                        });

    const {asFragment, findAllByText} = render(
        <Report defaultDate={new Date("2019-04-02") } />
    )

    await waitFor(() => {expect(mockReport.isDone()).toBeTruthy()})

    const items = await screen.findAllByText(/103/)
    expect(items).toHaveLength(1)
});
