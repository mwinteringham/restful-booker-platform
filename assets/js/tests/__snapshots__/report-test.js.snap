// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Multiple reports can be created in the Report component 1`] = `
<DocumentFragment>
  <div>
    <div
      class="rbc-calendar"
      style="height: 75vh;"
    >
      <div
        class="rbc-toolbar"
      >
        <span
          class="rbc-btn-group"
        >
          <button
            type="button"
          >
            Today
          </button>
          <button
            type="button"
          >
            Back
          </button>
          <button
            type="button"
          >
            Next
          </button>
        </span>
        <span
          class="rbc-toolbar-label"
        >
          April 2019
        </span>
        <span
          class="rbc-btn-group"
        />
      </div>
      <div
        aria-label="Month View"
        class="rbc-month-view"
        role="table"
      >
        <div
          class="rbc-row rbc-month-header"
          role="row"
        >
          <div
            class="rbc-header"
          >
            <span
              aria-sort="none"
              role="columnheader"
            >
              Sun
            </span>
          </div>
          <div
            class="rbc-header"
          >
            <span
              aria-sort="none"
              role="columnheader"
            >
              Mon
            </span>
          </div>
          <div
            class="rbc-header"
          >
            <span
              aria-sort="none"
              role="columnheader"
            >
              Tue
            </span>
          </div>
          <div
            class="rbc-header"
          >
            <span
              aria-sort="none"
              role="columnheader"
            >
              Wed
            </span>
          </div>
          <div
            class="rbc-header"
          >
            <span
              aria-sort="none"
              role="columnheader"
            >
              Thu
            </span>
          </div>
          <div
            class="rbc-header"
          >
            <span
              aria-sort="none"
              role="columnheader"
            >
              Fri
            </span>
          </div>
          <div
            class="rbc-header"
          >
            <span
              aria-sort="none"
              role="columnheader"
            >
              Sat
            </span>
          </div>
        </div>
        <div
          class="rbc-month-row"
          role="rowgroup"
        >
          <div
            class="rbc-row-bg"
          >
            <div
              class="rbc-day-bg rbc-off-range-bg"
            />
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
          </div>
          <div
            class="rbc-row-content"
            role="row"
          >
            <div
              class="rbc-row "
            >
              <div
                class="rbc-date-cell rbc-off-range"
                role="cell"
              >
                <button
                  class="rbc-button-link"
                  role="cell"
                  type="button"
                >
                  31
                </button>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <button
                  class="rbc-button-link"
                  role="cell"
                  type="button"
                >
                  01
                </button>
              </div>
              <div
                class="rbc-date-cell rbc-current"
                role="cell"
              >
                <button
                  class="rbc-button-link"
                  role="cell"
                  type="button"
                >
                  02
                </button>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <button
                  class="rbc-button-link"
                  role="cell"
                  type="button"
                >
                  03
                </button>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <button
                  class="rbc-button-link"
                  role="cell"
                  type="button"
                >
                  04
                </button>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <button
                  class="rbc-button-link"
                  role="cell"
                  type="button"
                >
                  05
                </button>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <button
                  class="rbc-button-link"
                  role="cell"
                  type="button"
                >
                  06
                </button>
              </div>
            </div>
            <div
              class="rbc-row"
            >
              <div
                class="rbc-row-segment"
                style="flex-basis: 14.285714285714285%; max-width: 14.285714285714285%;"
              >
                 
              </div>
              <div
                class="rbc-row-segment"
                style="flex-basis: 28.57142857142857%; max-width: 28.57142857142857%;"
              >
                <div
                  class="rbc-event rbc-event-allday"
                  tabindex="0"
                >
                  <div
                    class="rbc-event-content"
                    title="101"
                  >
                    101
                  </div>
                </div>
              </div>
            </div>
            <div
              class="rbc-row"
            >
              <div
                class="rbc-row-segment"
                style="flex-basis: 28.57142857142857%; max-width: 28.57142857142857%;"
              >
                 
              </div>
              <div
                class="rbc-row-segment"
                style="flex-basis: 28.57142857142857%; max-width: 28.57142857142857%;"
              >
                <div
                  class="rbc-event rbc-event-allday"
                  tabindex="0"
                >
                  <div
                    class="rbc-event-content"
                    title="102"
                  >
                    102
                  </div>
                </div>
              </div>
            </div>
            <div
              class="rbc-row"
            >
              <div
                class="rbc-row-segment"
                style="flex-basis: 28.57142857142857%; max-width: 28.57142857142857%;"
              >
                 
              </div>
              <div
                class="rbc-row-segment"
                style="flex-basis: 28.57142857142857%; max-width: 28.57142857142857%;"
              >
                <div
                  class="rbc-event rbc-event-allday"
                  tabindex="0"
                >
                  <div
                    class="rbc-event-content"
                    title="103"
                  >
                    103
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="rbc-month-row"
          role="rowgroup"
        >
          <div
            class="rbc-row-bg"
          >
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
          </div>
          <div
            class="rbc-row-content"
            role="row"
          >
            <div
              class="rbc-row "
            >
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <button
                  class="rbc-button-link"
                  role="cell"
                  type="button"
                >
                  07
                </button>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <button
                  class="rbc-button-link"
                  role="cell"
                  type="button"
                >
                  08
                </button>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <button
                  class="rbc-button-link"
                  role="cell"
                  type="button"
                >
                  09
                </button>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <button
                  class="rbc-button-link"
                  role="cell"
                  type="button"
                >
                  10
                </button>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <button
                  class="rbc-button-link"
                  role="cell"
                  type="button"
                >
                  11
                </button>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <button
                  class="rbc-button-link"
                  role="cell"
                  type="button"
                >
                  12
                </button>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <button
                  class="rbc-button-link"
                  role="cell"
                  type="button"
                >
                  13
                </button>
              </div>
            </div>
          </div>
        </div>
        <div
          class="rbc-month-row"
          role="rowgroup"
        >
          <div
            class="rbc-row-bg"
          >
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
          </div>
          <div
            class="rbc-row-content"
            role="row"
          >
            <div
              class="rbc-row "
            >
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <button
                  class="rbc-button-link"
                  role="cell"
                  type="button"
                >
                  14
                </button>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <button
                  class="rbc-button-link"
                  role="cell"
                  type="button"
                >
                  15
                </button>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <button
                  class="rbc-button-link"
                  role="cell"
                  type="button"
                >
                  16
                </button>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <button
                  class="rbc-button-link"
                  role="cell"
                  type="button"
                >
                  17
                </button>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <button
                  class="rbc-button-link"
                  role="cell"
                  type="button"
                >
                  18
                </button>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <button
                  class="rbc-button-link"
                  role="cell"
                  type="button"
                >
                  19
                </button>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <button
                  class="rbc-button-link"
                  role="cell"
                  type="button"
                >
                  20
                </button>
              </div>
            </div>
          </div>
        </div>
        <div
          class="rbc-month-row"
          role="rowgroup"
        >
          <div
            class="rbc-row-bg"
          >
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
          </div>
          <div
            class="rbc-row-content"
            role="row"
          >
            <div
              class="rbc-row "
            >
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <button
                  class="rbc-button-link"
                  role="cell"
                  type="button"
                >
                  21
                </button>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <button
                  class="rbc-button-link"
                  role="cell"
                  type="button"
                >
                  22
                </button>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <button
                  class="rbc-button-link"
                  role="cell"
                  type="button"
                >
                  23
                </button>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <button
                  class="rbc-button-link"
                  role="cell"
                  type="button"
                >
                  24
                </button>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <button
                  class="rbc-button-link"
                  role="cell"
                  type="button"
                >
                  25
                </button>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <button
                  class="rbc-button-link"
                  role="cell"
                  type="button"
                >
                  26
                </button>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <button
                  class="rbc-button-link"
                  role="cell"
                  type="button"
                >
                  27
                </button>
              </div>
            </div>
          </div>
        </div>
        <div
          class="rbc-month-row"
          role="rowgroup"
        >
          <div
            class="rbc-row-bg"
          >
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg"
            />
            <div
              class="rbc-day-bg rbc-off-range-bg"
            />
            <div
              class="rbc-day-bg rbc-off-range-bg"
            />
            <div
              class="rbc-day-bg rbc-off-range-bg"
            />
            <div
              class="rbc-day-bg rbc-off-range-bg"
            />
          </div>
          <div
            class="rbc-row-content"
            role="row"
          >
            <div
              class="rbc-row "
            >
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <button
                  class="rbc-button-link"
                  role="cell"
                  type="button"
                >
                  28
                </button>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <button
                  class="rbc-button-link"
                  role="cell"
                  type="button"
                >
                  29
                </button>
              </div>
              <div
                class="rbc-date-cell"
                role="cell"
              >
                <button
                  class="rbc-button-link"
                  role="cell"
                  type="button"
                >
                  30
                </button>
              </div>
              <div
                class="rbc-date-cell rbc-off-range"
                role="cell"
              >
                <button
                  class="rbc-button-link"
                  role="cell"
                  type="button"
                >
                  01
                </button>
              </div>
              <div
                class="rbc-date-cell rbc-off-range"
                role="cell"
              >
                <button
                  class="rbc-button-link"
                  role="cell"
                  type="button"
                >
                  02
                </button>
              </div>
              <div
                class="rbc-date-cell rbc-off-range"
                role="cell"
              >
                <button
                  class="rbc-button-link"
                  role="cell"
                  type="button"
                >
                  03
                </button>
              </div>
              <div
                class="rbc-date-cell rbc-off-range"
                role="cell"
              >
                <button
                  class="rbc-button-link"
                  role="cell"
                  type="button"
                >
                  04
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
